### PR TITLE
Add guards for unsafe dereferencing (ref #2687)

### PR DESCRIPTION
**Pull request by jacobq**
* In drawData
* In supplyDefaultsUpdateCalc

### CODE REVIEW

**Comment by jacobq** [comment]
OK, I think this is ready for review (I'm done doing `git push -f` here ;))

**Comment by etpinard** [comment]
Thanks very much for the PR!

I tested out https://codepen.io/watashiku/pen/pKJJNW?editors=1010 (from https://github.com/plotly/plotly.js/issues/2687#issue-328179735) off your branch and your patch didn't appear to fix the issues reported, it only hides the console errors.

Moreover, your patch made a few of tests fail.

I hope I'm wrong, but I think fixing the issues of https://github.com/plotly/plotly.js/issues/2687, #2643, and #2644 will have to wait a multi-week effort to implement https://github.com/plotly/plotly.js/issues/2644#issuecomment-391152782

File: src/plots/plots.js [hunk 1]
```diff
@@ -484,7 +484,7 @@ plots.supplyDefaults = function(gd, opts) {
 plots.supplyDefaultsUpdateCalc = function(oldCalcdata, newFullData) {
     for(var i = 0; i < newFullData.length; i++) {
         var newTrace = newFullData[i];
-        var cd0 = oldCalcdata[i][0];
+        var cd0 = oldCalcdata && oldCalcdata[i] && !oldCalcdata[i][0];
```
Comment (jacobq) [inline src/plots/plots.js]: Eek, I just noticed a really bad typo here: there shouldn't be a `!`

**Comment by jacobq** [comment]
@etpinard Please check/re-run that latest CI build. It looks to me like the failures aren't being caused by the changes in this PR (got stuck on "install dependencies" step). I'm curious to see if the test failures you saw earlier were because of my typo (had accidentally put a `!` in part of the sanity checking code, rendering it completely wrong).

You are right that this PR doesn't resolve the larger issue. It's just intended to prevent uncaught exceptions.

**Comment by etpinard** [comment]
> It's just intended to prevent uncaught exceptions.

... and that's why I'd vote :-1: on merging this patch.

**Comment by jacobq** [comment]
If you don't want to merge it that's fine, but I don't understand how broken code is better than gracefully broken code.

**Comment by etpinard** [comment]
> but I don't understand how broken code is better than gracefully failing code.

Because I don't think adding fallbacks to these broken things doesn't make plotly.js more graceful in my mind. I could be wrong though.

**Pull request closed without merging**